openpectus/protocol/engine_dispatcher.py

```diff
@@ -0,0 +1,122 @@
+
+
+    def check_aggregator_alive(self, aggregator_host: str):
+        aggregator_health_url = f"http://{aggregator_host}{AGGREGATOR_HEALTH_PATH}"
+        try:
+            resp = httpx.get(aggregator_health_url)
+        except httpx.ConnectError as ex:
+            print("Connection to Aggregator health end point failed.")
+            print(f"Status url: {aggregator_health_url}")
+            print(f"Error: {ex}")
+            print("OpenPectus Engine cannot start.")
+            exit(1)
+        if resp.is_error:
+            print("Aggregator health end point returned an unsuccessful result.")
+            print(f"Status url: {aggregator_health_url}")
+            print(f"HTTP status code returned: {resp.status_code}")
+            print("OpenPectus Engine cannot start.")
+            exit(1)
+
+    def __init__(self,  aggregator_host: str, uod_name: str) -> None:
+        super().__init__()
+        self._handlers: Dict[str, MessageHandler] = {}
+        self._engine_id = None
+
+        # TODO consider https/wss
+        self.post_url = f"http://{aggregator_host}{AGGREGATOR_REST_PATH}"
+        asyncio.create_task(self.setup_websocket(aggregator_host, uod_name))
+
+    async def setup_websocket(self, aggregator_host: str, uod_name: str):
+        self.check_aggregator_alive(aggregator_host)
+        while self._engine_id is None:
+            self._engine_id = await self.register_for_engine_id(uod_name)
+
+        rpc_url = f"ws://{aggregator_host}{AGGREGATOR_RPC_WS_PATH}"
+        rpc_methods = EngineDispatcher.EngineRpcMethods(self, self._engine_id)
+        self.rpc_client = WebSocketRpcClient(uri=rpc_url, methods=rpc_methods)
+        atexit.register(self.disconnect)
+        await self.rpc_client.__aenter__()
+
+    async def disconnect(self):
+        await self.rpc_client.__aexit__()
+
+    def post(self, message: EM.EngineMessage | EM.RegisterEngineMsg) -> M.MessageBase:
+        """ Send message via HTTP POST. """
+        if(not isinstance(message, EM.RegisterEngineMsg)): # Special case for registering
+            if(self._engine_id is None): return M.ErrorMessage(message="Engine did not have engine_id yet")
+            message.engine_id = self._engine_id
+        message_json = serialize(message)
+        response = requests.post(url=self.post_url, json=message_json)
+        if response.status_code == 200:
+            response_json = response.json()
+            value = deserialize(response_json)
+            return value
+        else:
+            message_type = type(message)
+            logger.error(f"Non-success http status code: {response.status_code} for input message type: {message_type}")
+            return M.ErrorMessage(message=f"Post failed with message {message_type}")
+
+    async def _dispatch_message(self, message: M.MessageBase):
+        """ Dispath message to registered handler. """
+        message_type = type(message).__name__
+        if message_type in self._handlers.keys():
```

Comment:
We need to handle both cases: 1) no handler registered and 2) handler failed. It is fine to add a TODO for now